File: core/nodesDev/gfToolsNodes/src/n_gfRigPSD.cpp
#include "../headers/n_gfRigPSD.h"

#define INPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(true);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(true);	\
	FNATTR.setKeyable(true);	\

#define OUTPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(false);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(false);	\
	FNATTR.setKeyable(false);	\



SphericalPSD::SphericalPSD() {}

SphericalPSD::~SphericalPSD() {}


MObject SphericalPSD::inBase;
MObject SphericalPSD::inPose;
MObject SphericalPSD::inTarget;
MObject SphericalPSD::inTargetEnvelope;
MObject SphericalPSD::inTargetFalloff;
MObject SphericalPSD::outWeight;


void * SphericalPSD::creator() {
	return new SphericalPSD();
}


MStatus SphericalPSD::initialize() {
	MStatus status = MStatus::kFailure;

	MFnNumericAttribute nAttr;
	MFnMatrixAttribute mAttr;

	inBase = mAttr.create("baseMtx", "bmtx", MFnMatrixAttribute::kDouble, &status);
	INPUT_ATTR(mAttr);

	inPose = mAttr.create("poseMtx", "pmtx", MFnMatrixAttribute::kDouble, &status);
	INPUT_ATTR(mAttr);

	inTarget = mAttr.create("targetMtx", "tmtx", MFnMatrixAttribute::kDouble, &status);
	mAttr.setArray(true);
	INPUT_ATTR(mAttr);

	inTargetEnvelope = nAttr.create("targetEnvelope", "te", MFnNumericData::kDouble, 1.0, &status);
	nAttr.setMin(0.0);
	nAttr.setMax(1.0);
	nAttr.setArray(true);
	INPUT_ATTR(nAttr);

	inTargetFalloff = nAttr.create("targetFalloff", "tf", MFnNumericData::kDouble, 45.0, &status);
	nAttr.setMin(0.0);
	nAttr.setSoftMax(135.0);
	nAttr.setMax(180.0);
	nAttr.setArray(true);
	INPUT_ATTR(nAttr);

	outWeight = nAttr.create("outWeights", "ow", MFnNumericData::kDouble, 0.0, &status);
	nAttr.setMin(0.0);
	nAttr.setMax(1.0);
	nAttr.setArray(true);
	OUTPUT_ATTR(nAttr);

	addAttribute(inBase);
	addAttribute(inPose);
	addAttribute(inTarget);
	addAttribute(inTargetEnvelope);
	addAttribute(inTargetFalloff);
	addAttribute(outWeight);

	attributeAffects(inBase, outWeight);
	attributeAffects(inPose, outWeight);
	attributeAffects(inTarget, outWeight);
	attributeAffects(inTargetEnvelope, outWeight);
	attributeAffects(inTargetFalloff, outWeight);

	return status;
}


MStatus SphericalPSD::compute(const MPlug & plug, MDataBlock & dataBlock) {
	if (plug != outWeight) {
		return MStatus::kUnknownParameter;
	}

	MMatrix baseMtx = dataBlock.inputValue(inBase).asMatrix();
	MMatrix poseMtx = dataBlock.inputValue(inPose).asMatrix();
	MArrayDataHandle targetHandle = dataBlock.inputArrayValue(inTarget);
	MArrayDataHandle envelopeHandle = dataBlock.inputArrayValue(inTargetEnvelope);
	MArrayDataHandle falloffHandle = dataBlock.inputArrayValue(inTargetFalloff);
	MArrayDataHandle weightsHandle = dataBlock.outputArrayValue(outWeight);

	MTransformationMatrix mtxFn(baseMtx);
	MVector vBase = mtxFn.translation(MSpace::kWorld);
	mtxFn = MTransformationMatrix(poseMtx);
	MVector vPose = mtxFn.translation(MSpace::kWorld);

	MVector vCurPose = vPose - vBase;
	MVector nCurPose = vCurPose.normal();

	std::vector<MVector> targetPoseList;
	std::vector<double> envelopeList;
	std::vector<double> falloffList;

	uint32_t index = targetHandle.elementCount();
	for (uint32_t i = 0; i < index; i++) {
		targetHandle.jumpToArrayElement(i);
		MMatrix mtx(targetHandle.inputValue().asMatrix());
		mtxFn = MTransformationMatrix(mtx);
		MVector vector(mtxFn.translation(MSpace::kWorld));
		MVector vTargetPose = vector - vBase;
		MVector nTargetPose = vTargetPose.normal();
		targetPoseList.push_back(nTargetPose);
	}

	index = envelopeHandle.elementCount();
	for (uint32_t i = 0; i < index; i++) {
		envelopeHandle.jumpToArrayElement(i);
		double env = envelopeHandle.inputValue().asDouble();
		envelopeList.push_back(env);
	}

	index = falloffHandle.elementCount();
	for (uint32_t i = 0; i < index; i++) {
		falloffHandle.jumpToArrayElement(i);
		double falloff = falloffHandle.inputValue().asDouble() * M_PI / 180.0;
		falloffList.push_back(falloff);
	}

	for (uint32_t i = 0; i < weightsHandle.elementCount(); i++) {
		weightsHandle.jumpToArrayElement(i);
		MDataHandle resultHandle = weightsHandle.outputValue();

		if ((i < weightsHandle.elementCount()) &&
			(i < targetPoseList.size()) &&
			(i < falloffList.size()) &&
			(i < envelopeList.size())) {

			double alpha = std::acos(targetPoseList[i] * nCurPose);
			double ratio = std::min(std::max(alpha / falloffList[i], -1.0), 1.0);

			double resultWeight;
			if (ratio == 0.0) {
				resultWeight = envelopeList[i] * 1.0;
			}
			else if (ratio > 0.0) {
				resultWeight = envelopeList[i] * (1.0 - ratio);
			}
			else if (ratio < 0.0) {
				resultWeight = envelopeList[i] * (1.0 + ratio);
			}

			resultHandle.setDouble(resultWeight);
		}
		else {
			resultHandle.setDouble(0.0);
		}
		
	}

	weightsHandle.setAllClean();

	return MStatus::kSuccess;
}

File: core/nodesDev/gfToolsNodes/py/n_gfRigPSD.py
# -*- coding: utf-8 -*-
"""
Copyright (c) 2019 Giuliano França

Redistribution:
    Something here.

Maya Node:
    [This is a prototype version of the gfRigIKVChain node. You should be using the related C++ version.]
    This node is a test node and only performs test operations with one input value.
    https://www.desmos.com/calculator/nfggjvzpkn

Requirements:
    Maya 2017 or above.

Todo:
    * NDA

This code supports Pylint. Rc file in project.
"""
# pylint: disable=E0401
# E0401 = Supress Maya modules import error

import math
import maya.api.OpenMaya as om2


def maya_useNewAPI():
    """ Function to Maya recognize the use of the Python API 2.0. """
    # pylint: disable=C0103, w0107
    pass


def INPUT_ATTR(FNATTR):
    """ Configure a input attribute. """
    # pylint: disable=C0103
    FNATTR.writable = True
    FNATTR.readable = True
    FNATTR.storable = True
    FNATTR.keyable = True


def OUTPUT_ATTR(FNATTR):
    """ Configure a output attribute. """
    # pylint: disable=C0103
    FNATTR.writable = False
    FNATTR.readable = True
    FNATTR.storable = False
    FNATTR.keyable = False


class SphericalPSD(om2.MPxNode):
    """ Main class of gfRigPSD node. """

    kNODE_NAME = ""
    kNODE_CLASSIFY = ""
    kNODE_ID = ""

    inBase = om2.MObject()
    inPose = om2.MObject()
    inTarget = om2.MObject()
    inTargetEnvelope = om2.MObject()
    inTargetFalloff = om2.MObject()
    outWeight = om2.MObject()

    def __init__(self):
        """ Constructor. """
        om2.MPxNode.__init__(self)

    @staticmethod
    def creator():
        """ Maya creator function. """
        return SphericalPSD()

    @staticmethod
    def initialize():
        """
        Defines the set of attributes for this node. The attributes declared in this function are assigned
        as static members to IKVChain class. Instances of IKVChain will use these attributes to create plugs
        for use in the compute() method.
        """
        nAttr = om2.MFnNumericAttribute()
        mAttr = om2.MFnMatrixAttribute()

        SphericalPSD.inBase = mAttr.create("base", "b", om2.MFnMatrixAttribute.kDouble)
        INPUT_ATTR(mAttr)

        SphericalPSD.inPose = mAttr.create("pose", "p", om2.MFnMatrixAttribute.kDouble)
        INPUT_ATTR(mAttr)

        SphericalPSD.inTarget = mAttr.create("target", "t", om2.MFnMatrixAttribute.kDouble)
        mAttr.array = True
        INPUT_ATTR(mAttr)

        SphericalPSD.inTargetEnvelope = nAttr.create("targetEnvelope", "te", om2.MFnNumericData.kFloat, 1.0)
        nAttr.setMin(0.0)
        nAttr.setMax(1.0)
        nAttr.array = True
        INPUT_ATTR(nAttr)

        SphericalPSD.inTargetFalloff = nAttr.create("targetFalloff", "tf", om2.MFnNumericData.kFloat, 45.0)
        nAttr.setMin(0.0)
        nAttr.setSoftMax(135.0)
        nAttr.setMax(180.0)
        nAttr.array = True
        INPUT_ATTR(nAttr)

        SphericalPSD.outWeight = nAttr.create("outWeight", "ow", om2.MFnNumericData.kFloat, 0.0)
        nAttr.setMin(0.0)
        nAttr.setMax(1.0)
        nAttr.array = True
        OUTPUT_ATTR(nAttr)

        SphericalPSD.addAttribute(SphericalPSD.inBase)
        SphericalPSD.addAttribute(SphericalPSD.inPose)
        SphericalPSD.addAttribute(SphericalPSD.inTarget)
        SphericalPSD.addAttribute(SphericalPSD.inTargetEnvelope)
        SphericalPSD.addAttribute(SphericalPSD.inTargetFalloff)
        SphericalPSD.addAttribute(SphericalPSD.outWeight)

        SphericalPSD.attributeAffects(SphericalPSD.inBase, SphericalPSD.outWeight)
        SphericalPSD.attributeAffects(SphericalPSD.inPose, SphericalPSD.outWeight)
        SphericalPSD.attributeAffects(SphericalPSD.inTarget, SphericalPSD.outWeight)
        SphericalPSD.attributeAffects(SphericalPSD.inTargetEnvelope, SphericalPSD.outWeight)
        SphericalPSD.attributeAffects(SphericalPSD.inTargetFalloff, SphericalPSD.outWeight)

    def compute(self, plug, dataBlock):
        """
        Node computation method:
            * plug is a connection point related to one of our node attributes (either an input or an output).
            * dataBlock contains the data on which we will base our computations.
        """
        # pylint: disable=R0201
        if plug == SphericalPSD.outWeight:
            baseMtx = dataBlock.inputValue(SphericalPSD.inBase).asMatrix()
            poseMtx = dataBlock.inputValue(SphericalPSD.inPose).asMatrix()
            targetHandle = dataBlock.inputArrayValue(SphericalPSD.inTarget)
            envelopeHandle = dataBlock.inputArrayValue(SphericalPSD.inTargetEnvelope)
            falloffHandle = dataBlock.inputArrayValue(SphericalPSD.inTargetFalloff)
            weightsHandle = dataBlock.outputArrayValue(SphericalPSD.outWeight)

            mtxFn = om2.MTransformationMatrix(baseMtx)
            vBase = mtxFn.translation(om2.MSpace.kWorld)
            mtxFn = om2.MTransformationMatrix(poseMtx)
            vPose = mtxFn.translation(om2.MSpace.kWorld)

            vCurPose = vPose - vBase
            nCurPose = vCurPose.normal()

            targetPoseList = []
            envelopeList = []
            falloffList = []

            index = len(targetHandle)
            for i in range(index):
                targetHandle.jumpToLogicalElement(i)
                mtx = targetHandle.inputValue().asMatrix()
                mtxFn = om2.MTransformationMatrix(mtx)
                vector = mtxFn.translation(om2.MSpace.kWorld)
                vTargetPose = vector - vBase
                nTargetPose = vTargetPose.normal()
                targetPoseList.append(nTargetPose)

            index = len(envelopeHandle)
            for i in range(index):
                envelopeHandle.jumpToLogicalElement(i)
                env = envelopeHandle.inputValue().asFloat()
                envelopeList.append(env)

            index = len(falloffHandle)
            for i in range(index):
                falloffHandle.jumpToLogicalElement(i)
                falloff = math.radians(falloffHandle.inputValue().asFloat())
                falloffList.append(falloff)

            for i in range(len(weightsHandle)):
                weightsHandle.jumpToLogicalElement(i)
                resultHandle = weightsHandle.outputValue()
                if (i < len(weightsHandle) and
                        i < len(targetPoseList) and
                        i < len(falloffList) and
                        i < len(envelopeList)):
                    alpha = math.acos(targetPoseList[i] * nCurPose)
                    ratio = min(max(alpha/falloffList[i], -1.0), 1.0)
                    if ratio == 0.0:
                        resultWeight = envelopeList[i] * 1.0
                    elif ratio > 0.0:
                        resultWeight = envelopeList[i] * (1.0 - ratio)
                    elif ratio < 0.0:
                        resultWeight = envelopeList[i] * (1.0 + ratio)

                    resultHandle.setFloat(resultWeight)
                else:
                    resultHandle.setFloat(0.0)

            weightsHandle.setAllClean()
